File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
private:
    ListNode* head_ = nullptr;
    ListNode* tail_ = nullptr;
    void insert(int val){
        ListNode* nnode = new ListNode(val);
        if(!head_){
            head_ = tail_ = nnode;
        }
        else{
            tail_->next = nnode;
            tail_ = nnode;
        }
    }

public:
    void helper(ListNode* head){
        if(head == nullptr){
            return;
        }
        helper(head->next);
        insert(head->val);
    }
    ListNode* reverseList(ListNode* head) {
        helper(head);
        return head_;
    }
};

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        """
            [3,null]

            mp
            3 -> Node(3) [3,null]
            7 -> Node(7) [7,3]
            4 -> Node(4)

        """
        self.head, self.tail = None, None
        mp = {}

        if not head:
            return None

        def insert(curr):
            if self.head == self.tail == None:
                self.head = self.tail = curr 
            
            else:
                self.tail.next = curr
                self.tail = curr

        curr = head
        while curr:
            if curr not in mp:
                nnode = Node(curr.val)
                mp[curr] = nnode
            
            if curr.random and curr.random not in mp:
                nnode = Node(curr.random.val)
                mp[curr.random] = nnode 
            
            if curr.random:
                mp[curr].random = mp[curr.random]

            insert(mp[curr])
            curr = curr.next

        return mp[head]
